tests/kernel/interrupt/src/shared_irq.c

```diff
@@ -0,0 +1,337 @@
+{
+	int irq_line_1, irq_line_2, irq_priority, i;
+
+#if defined(CONFIG_GIC)
+	irq_line_1 = GIC_IRQ_LINE_1;
+	irq_line_2 = GIC_IRQ_LINE_2;
+	irq_priority = GIC_IRQ_PRIORITY;
+#elif defined(CONFIG_CPU_CORTEX_M)
+	irq_line_1 = get_available_nvic_line(CONFIG_NUM_IRQS);
+	irq_line_2 = get_available_nvic_line(irq_line_1);
+	irq_priority = 1;
+#else
+#error "Unsupported architecture."
+#endif
+
+	arch_irq_connect_dynamic(irq_line_1, irq_priority, test_isr_0, 0, 0);
+
+	if (_sw_isr_table[irq_line_1].isr != test_isr_0) {
+		return false;
+	}
+	if (_sw_isr_table[irq_line_1].arg) {
+		return false;
+	}
+	if (_shared_irq_table[irq_line_1].client_num) {
+		return false;
+	}
+
+	arch_irq_connect_dynamic(irq_line_1, irq_priority, test_isr_1, (void *)1, 0);
+
+	if (_sw_isr_table[irq_line_1].isr != shared_isr) {
+		return false;
+	}
+	if (_sw_isr_table[irq_line_1].arg != &_shared_irq_table[irq_line_1]) {
+		return false;
+	}
+	if (_shared_irq_table[irq_line_1].client_num != 2) {
+		return false;
+	}
+
+	/* here it's mandatory that test_isr_0/NULL is the first client, while
+	 * test_isr_1/1 is the second client.
+	 */
+	if (!client_exists_at_index(test_isr_0, 0, irq_line_1, 0)) {
+		return 0;
+	}
+	if (!client_exists_at_index(test_isr_1, (void *)1, irq_line_1, 1)) {
+		return 0;
+	}
+
+	arch_irq_connect_dynamic(irq_line_2, irq_priority, test_isr_2, (void *)2, 0);
+
+	if (_sw_isr_table[irq_line_2].isr != test_isr_2) {
+		return false;
+	}
+	if (_sw_isr_table[irq_line_2].arg != (void *)2) {
+		return false;
+	}
+	if (_shared_irq_table[irq_line_2].client_num) {
+		return false;
+	}
+
+	irq_enable(irq_line_1);
+	irq_enable(irq_line_2);
+
+	trigger_irq(irq_line_1);
+	trigger_irq(irq_line_2);
+
+	/* wait 5ms before checking the results */
+	k_busy_wait(5000);
+
+	for (i = 0; i < TEST_VECTOR_SIZE; i++) {
+		if (test_vector[i] != result_vector[i]) {
+			return false;
+		}
+	}
+
+	irq_disable(irq_line_1);
+	irq_disable(irq_line_2);
+
+	reset_test_vector();
+
+	/* remove test_isr_0/NULL pair. After this statement we expect
+	 * irq_line_1 to be unshared.
+	 */
+	arch_irq_disconnect_dynamic(irq_line_1, irq_priority, test_isr_0, 0, 0);
+
+	if (_sw_isr_table[irq_line_1].isr != test_isr_1) {
+		return false;
+	}
+	if (_sw_isr_table[irq_line_1].arg != (void *)1) {
+		return false;
+	}
+	if (_shared_irq_table[irq_line_1].client_num) {
+		return false;
+	}
+
+	irq_enable(irq_line_1);
+	trigger_irq(irq_line_1);
+
+	/* wait 5ms before checking the results */
+	k_busy_wait(5000);
+
+	for (i = 0; i < TEST_VECTOR_SIZE; i++) {
+		if (i == 1) {
+			if (test_vector[i] != result_vector[i]) {
+				return false;
+			}
+			continue;
+		}
+
+		if (test_vector[i]) {
+			return false;
+		}
+	}
+
+	irq_disable(irq_line_1);
+
+	reset_test_vector();
+
+	/* pass */
+	return true;
+}
+
+ZTEST(interrupt_feature, test_shared_irq)
+{
+	/* TODO: make this test work on other architectures as well.
+	 * What are the equivalents of the SGIs on the other architectures?
+	 */
```

Comment:
what do you mean? `trigger_irq` should support basically all the architectures.

Reply:
Removed comment altogether as it was misleading. It was mostly referring to which interrupt lines to use on other architectures such that the corresponding entry in `_sw_isr_table` would be `NULL/z_irq_spurious`.